panic with empty handler func

Panic occurs if an empty handler that does not write a status code is called more than once.

```go
package stampede_test

import (
	"net/http"
	"net/http/httptest"
	"testing"
	"time"

	"github.com/go-chi/stampede"
)

func TestPanic(t *testing.T) {
	mux := http.NewServeMux()
	middleware := stampede.Handler(100, 1*time.Hour)
	mux.Handle("/", middleware(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
		t.Log(r.Method, r.URL)
	})))

	ts := httptest.NewServer(mux)
	defer ts.Close()

	{
		req, err := http.NewRequest(http.MethodGet, ts.URL, nil)
		if err != nil {
			t.Fatal(err)
		}
		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			t.Fatal(err)
		}
		defer resp.Body.Close()
		t.Log(resp.StatusCode)
	}
	{
		req, err := http.NewRequest(http.MethodGet, ts.URL, nil)
		if err != nil {
			t.Fatal(err)
		}
		resp, err := http.DefaultClient.Do(req)
		if err != nil {
			t.Fatal(err)
		}
		defer resp.Body.Close()
		t.Log(resp.StatusCode)
	}
}
```

```
2021/03/05 10:52:17 http: panic serving 127.0.0.1:59282: invalid WriteHeader code 0
goroutine 34 [running]:
net/http.(*conn).serve.func1(0xc000208000)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:1824 +0x153
panic(0x744ac0, 0xc000224100)
        /home/heijo/ghq/go.googlesource.com/go/src/runtime/panic.go:980 +0x4d7
net/http.checkWriteHeaderCode(...)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:1112
net/http.(*response).WriteHeader(0xc0002380e0, 0x0)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:1146 +0x72b
github.com/go-chi/stampede.HandlerWithKey.func1.1(0x82d310, 0xc0002380e0, 0xc00022c100)
        /home/heijo/ghq/github.com/go-chi/stampede/http.go:120 +0x4c3
net/http.HandlerFunc.ServeHTTP(0xc000222340, 0x82d310, 0xc0002380e0, 0xc00022c100)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:2069 +0x44
github.com/go-chi/stampede.Handler.func2.1(0x82d310, 0xc0002380e0, 0xc00022c100)
        /home/heijo/ghq/github.com/go-chi/stampede/http.go:49 +0x16d
net/http.HandlerFunc.ServeHTTP(0xc00011da10, 0x82d310, 0xc0002380e0, 0xc00022c100)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:2069 +0x44
net/http.(*ServeMux).ServeHTTP(0xc000168fc0, 0x82d310, 0xc0002380e0, 0xc00022c100)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:2448 +0x1ad
net/http.serverHandler.ServeHTTP(0xc0001e2000, 0x82d310, 0xc0002380e0, 0xc00022c100)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:2887 +0xa3
net/http.(*conn).serve(0xc000208000, 0x82d980, 0xc000222000)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:1952 +0x8cd
created by net/http.(*Server).Serve
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:3013 +0x398
2021/03/05 10:52:17 http: panic serving 127.0.0.1:59284: invalid WriteHeader code 0
goroutine 38 [running]:
net/http.(*conn).serve.func1(0xc000208140)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:1824 +0x153
panic(0x744ac0, 0xc00028e020)
        /home/heijo/ghq/go.googlesource.com/go/src/runtime/panic.go:980 +0x4d7
net/http.checkWriteHeaderCode(...)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:1112
net/http.(*response).WriteHeader(0xc000292000, 0x0)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:1146 +0x72b
github.com/go-chi/stampede.HandlerWithKey.func1.1(0x82d310, 0xc000292000, 0xc00022c300)
        /home/heijo/ghq/github.com/go-chi/stampede/http.go:120 +0x4c3
net/http.HandlerFunc.ServeHTTP(0xc00028a080, 0x82d310, 0xc000292000, 0xc00022c300)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:2069 +0x44
github.com/go-chi/stampede.Handler.func2.1(0x82d310, 0xc000292000, 0xc00022c300)
        /home/heijo/ghq/github.com/go-chi/stampede/http.go:49 +0x16d
net/http.HandlerFunc.ServeHTTP(0xc00011da10, 0x82d310, 0xc000292000, 0xc00022c300)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:2069 +0x44
net/http.(*ServeMux).ServeHTTP(0xc000168fc0, 0x82d310, 0xc000292000, 0xc00022c300)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:2448 +0x1ad
net/http.serverHandler.ServeHTTP(0xc0001e2000, 0x82d310, 0xc000292000, 0xc00022c300)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:2887 +0xa3
net/http.(*conn).serve(0xc000208140, 0x82d980, 0xc0002223c0)
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:1952 +0x8cd
created by net/http.(*Server).Serve
        /home/heijo/ghq/go.googlesource.com/go/src/net/http/server.go:3013 +0x398
--- FAIL: TestPanic (0.00s)
    http_test.go:16: GET /
    http_test.go:32: 200
    http_test.go:41: Get "http://127.0.0.1:43837": EOF
FAIL
FAIL    github.com/go-chi/stampede      0.004s
FAIL
```

--- a/http.go
+++ b/http.go
@@ -97,14 +97,14 @@ func HandlerWithKey(cacheSize int, ttl time.Duration, keyFunc ...func(r *http.Re
 
 				next.ServeHTTP(ww, r)
 
-				// the handler may not write header and body in some logic,
-				// while writing only the body, an attempt is made to write the default header (http.StatusOK)
-
 				val := responseValue{
 					headers: ww.Header(),
 					status:  ww.Status(),
 					body:    buf.Bytes(),
-					skip:    ww.IsHeaderWrong(),
+
+					// the handler may not write header and body in some logic,
+					// while writing only the body, an attempt is made to write the default header (http.StatusOK)
+					skip: ww.IsHeaderWrong(),
 				}
 				return val, nil
 			})
@@ -177,17 +177,6 @@ func (b *responseWriter) WriteHeader(code int) {
 }
 
 func (b *responseWriter) IsHeaderWrong() bool {
-	// Issue 22880: require valid WriteHeader status codes.
-	// For now we only enforce that it's three digits.
-	// In the future we might block things over 599 (600 and above aren't defined
-	// at https://httpwg.org/specs/rfc7231.html#status.codes)
-	// and we might block under 200 (once we have more mature 1xx support).
-	// But for now any three digits.
-	//
-	// We used to send "HTTP/1.1 000 0" on the wire in responses but there's
-	// no equivalent bogus thing we can realistically send in HTTP/2,
-	// so we'll consistently panic instead and help people find their bugs
-	// early. (We can't return an error from WriteHeader even if we wanted to.)
 	return !b.wroteHeader && (b.code < 100 || b.code > 999)
 }
 

--- a/http.go
+++ b/http.go
@@ -97,10 +97,14 @@ func HandlerWithKey(cacheSize int, ttl time.Duration, keyFunc ...func(r *http.Re
 
 				next.ServeHTTP(ww, r)
 
+				// the handler may not write header and body in some logic,
+				// while writing only the body, an attempt is made to write the default header (http.StatusOK)
+
 				val := responseValue{
 					headers: ww.Header(),
 					status:  ww.Status(),
 					body:    buf.Bytes(),
+					skip:    ww.IsHeaderWrong(),
 				}
 				return val, nil
 			})
@@ -119,6 +123,10 @@ func HandlerWithKey(cacheSize int, ttl time.Duration, keyFunc ...func(r *http.Re
 			resp, ok := val.(responseValue)
 			if !ok {
 				panic("stampede: handler received unexpected response value type")
+			}
+
+			if resp.skip {
+				return
 			}
 
 			header := w.Header()
@@ -149,6 +157,7 @@ type responseValue struct {
 	headers http.Header
 	status  int
 	body    []byte
+	skip    bool
 }
 
 type responseWriter struct {
@@ -165,6 +174,21 @@ func (b *responseWriter) WriteHeader(code int) {
 		b.wroteHeader = true
 		b.ResponseWriter.WriteHeader(code)
 	}
+}
+
+func (b *responseWriter) IsHeaderWrong() bool {
+	// Issue 22880: require valid WriteHeader status codes.
+	// For now we only enforce that it's three digits.
+	// In the future we might block things over 599 (600 and above aren't defined
+	// at https://httpwg.org/specs/rfc7231.html#status.codes)
+	// and we might block under 200 (once we have more mature 1xx support).
+	// But for now any three digits.
+	//
+	// We used to send "HTTP/1.1 000 0" on the wire in responses but there's
+	// no equivalent bogus thing we can realistically send in HTTP/2,
+	// so we'll consistently panic instead and help people find their bugs
+	// early. (We can't return an error from WriteHeader even if we wanted to.)
+	return !b.wroteHeader && (b.code < 100 || b.code > 999)
 }
 
 func (b *responseWriter) Write(buf []byte) (int, error) {

--- a/stampede_test.go
+++ b/stampede_test.go
@@ -248,3 +248,39 @@ func TestIssue6_BypassCORSHeaders(t *testing.T) {
 	// expect to have only one actual hit
 	assert.Equal(t, uint64(1), count)
 }
+
+func TestPanic(t *testing.T) {
+	mux := http.NewServeMux()
+	middleware := stampede.Handler(100, 1*time.Hour)
+	mux.Handle("/", middleware(http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		t.Log(r.Method, r.URL)
+	})))
+
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+
+	{
+		req, err := http.NewRequest(http.MethodGet, ts.URL, nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer resp.Body.Close()
+		t.Log(resp.StatusCode)
+	}
+	{
+		req, err := http.NewRequest(http.MethodGet, ts.URL, nil)
+		if err != nil {
+			t.Fatal(err)
+		}
+		resp, err := http.DefaultClient.Do(req)
+		if err != nil {
+			t.Fatal(err)
+		}
+		defer resp.Body.Close()
+		t.Log(resp.StatusCode)
+	}
+}

